Add quantile aggregator plugin (#8594)
influxdata/telegraf: plugins/aggregators/quantile/README.md

# Quantile Aggregator Plugin

The quantile aggregator plugin aggregates specified quantiles for each numeric field
per metric it sees and emits the quantiles every `period`.

### Configuration

```toml
[[aggregators.quantile]]
  ## General Aggregator Arguments:
  ## The period on which to flush & clear the aggregator.
  period = "30s"

  ## If true, the original metric will be dropped by the
  ## aggregator and will not get sent to the output plugins.
  drop_original = false

  ## Quantiles to output in the range [0,1]
  # quantiles = [0.25, 0.5, 0.75]

  ## Type of approximation
  ## Supported are:
  ##  "t-digest" -- approximation using centroids, can cope with large number of samples
  ##  "exact R7" -- exact approximation also used by Excel or NumPy (Hyndman & Fan 1996 R7)
  ##  "exact R8" -- exact approximation (Hyndman & Fan 1996 R8)
  ## NOTE: Do not use "exact" approximations with large number of samples
  ##       to not impair performance or memory consumption!

>   # approximator = "t-digest"

Reviewer: It doesn't make sense to me to call this setting "approximator" when two of the options are exact, not approximate.  What do you think about calling it algorithm instead?

Author: Will do.

  ## Compression for approximation (t-digest). The value needs to be
  ## greater or equal to 1.0. Smaller values will result in more
  ## performance but less accuracy.
  # compression = 100.0
```

#### Approximator types
##### t-digest
Proposed by [Dunning & Ertl (2019)][tdigest_paper] this type uses a
special data-structure to cluster data. These clusters are later used
to approximate the requested quantiles. The bounds of the approximation
can be controlled by the `compression` setting where smaller values
result in higher performance but less accuracy.

Due to its incremental nature, this approximatore can cope with large
numbers of samples and is therefor **recommended**.

> For implementation details see the underlying [a golang library][tdigest_lib].

Reviewer: ```suggestion
For implementation details see the underlying [golang library][tdigest_lib].
```

| exact R7      |            3  | 9782946 ns/op |
| exact R8      |            3  | 9158205 ns/op |
| t-digest      |          100  |  899204 ns/op |
| exact R7      |          100  | 7868816 ns/op |
| exact R8      |          100  | 8099612 ns/op |

### Measurements
Measurement names are passed trough this aggregator.

### Fields

For all numeric fields (int32/64, uint32/64 and float32/64) new *quantile*
fields are aggregated in the form `<fieldname>_<quantile*100>`. Other field
types (e.g. boolean, string) are ignored and dropped from the output.

For example passing in the following metric as *input*:
- somemetric
  - average_response_ms (float64)
  - minimum_response_ms (float64)
  - maximum_response_ms (float64)
  - status (string)
  - ok (boolean)

and the default setting for `quantiles ` you get the following *output*
- somemetric
  - average_response_ms_025 (float64)
  - average_response_ms_050 (float64)
  - average_response_ms_075 (float64)
  - minimum_response_ms_025 (float64)
  - minimum_response_ms_050 (float64)
  - minimum_response_ms_075 (float64)
  - maximum_response_ms_025 (float64)
  - maximum_response_ms_050 (float64)
  - maximum_response_ms_075 (float64)

> As you can see `status` and `ok` are dropped due to their type. Furthermore,
> the number of resulting fields scales with the number of `quantiles` specified.

Reviewer: ```suggestion
The `status` and `ok` fields are dropped because they are not numeric.  Note that the
number of resulting fields scales with the number of `quantiles` specified.
```
